### PR TITLE
Add English messages resource

### PR DESCRIPTION
**Pull request by tonygermano**
Fixes https://github.com/mozilla/rhino/issues/680

Adds blank English language file so that it can be selected as a language for the Context Locale when the JVM default Locale's language is not English, but is translated.

Message keys not translated in the language specific properties file will fall through to the base (no language) file.

### CODE REVIEW

**Comment by p-bakker** [comment]
Not sure this is the correct fix: for any Locale that doesn't use French (fr) as the language, the code `ResourceBundle.getBundle("org.mozilla.javascript.resources.Messages", locale);` should already pick up the `Messages.properties`

Adding an empty `Messages_en.properties` shouldn't make any difference.

My hunch is more that the reported issue has more to do with the code at https://github.com/mozilla/rhino/blob/master/src/org/mozilla/javascript/ScriptRuntime.java#L4510: if there's no current context, it uses the default Locale, which, in the case of the reporter of #680 is French

**Comment by tonygermano** [comment]
> for any Locale that doesn't use French (fr) as the language, the code ... should already pick up the `Messages.properties`

That is correct, but if more translations are added, then it would be true of any of them if they match the default Locale. Without this file, there is no way to explicitly select English. Attempting to select English when French is the default fails, because it is no different than any other language that does not have a translation. Currently English is no different than any other language that isn't translated, which the exception that the base file for untranslated messages happens to already be in English.

`defaultLocaleUsesOtherTranslatedLanguageAndContextLocaleUsesDefaultLanguage` is the only test that fails without the fix, and the only workaround is the one specified in the issue, where a local with no language is used at the Context level.

**Comment by p-bakker** [comment]
I now see, odd behavior of the ResourceBundle, that seems to go back to using Java on the DeskTop.

The way to fix it though seems to be `ResourceBundle.getBundle(RESOURCE_BUNDLE_NAME, locale, 
    ResourceBundle.Control.getNoFallbackControl(
        ResourceBundle.Control.FORMAT_DEFAULT));`, which results in ignoring the default Locale of the vm

**Comment by tonygermano** [comment]
I had originally implemented my own `Resource.Control` and overrode the `getFallBackLocale` method, but then I was running into issues where certain conditions would send it into an infinite loop (also having to do with the fact that it couldn't actually find an English translation.)

This solution seemed both elegant and simpler.

When a user's default Locale language is translated (only `fr` at this time) I believe it is the proper behavior that when an untranslated Locale is set on the Context, that it should fall back to the default Locale, not to English. In that case, English needs to be selectable as a language if that is what the user desires to see.

The "no language" implementation should be the sole message source only when both the default and context locales are untranslated.

The French file uses the same principle as the empty English file. For example, if you trigger one of the recently added messages, like trying to add a bigint and number together, it will display the (no language) English message, because the French file does not include an entry for that key, but the base file does. The ResourceBundle is set up as a hierarchy for that reason (you can stack these several layers deep by including country codes and modules in the file names.)

**Comment by p-bakker** [comment]
> When a user's default Locale language is translated (only fr at this time) I believe it is the proper behavior that when an untranslated Locale is set on the Context, that it should fall back to the default Locale

I think that is debatable, personally I rather ignore the VM's default Locale and it seems to me the OP of #680 expects that as well

> but then I was running into issues where certain conditions would send it into an infinite loop
Does that also happen when using `ResourceBundle.Control.getNoFallbackControl( ResourceBundle.Control.FORMAT_DEFAULT)`?

**Comment by tonygermano** [comment]
How I understood the OP's problem is that they were trying to select English as a language and could not, because technically we don't have an English translation. We only have a base translation that happens to be in English.

If I am a user in France seeing messages in French, and I set the context language to Spanish, when a Spanish translation does not exist, I would expect it to fall back to French, not English.

The context already returns the VM's default Locale when the Locale is not set.

https://github.com/mozilla/rhino/blob/76b37a1353b0f1b8bbe76b6e59b06878fee7f378/src/org/mozilla/javascript/Context.java#L740-L748

I guess that behavior is really what is in question, as that is what I was trying to achieve. I didn't think setting the context language to one that is not translated should be treated any differently than not setting a context language in the first place.

Another benefit of assuming this behavior is that this PR is a no-code change. It only adds a test and a blank file (ignoring comments.)

**Comment by tonygermano** [comment]
In other words, this is how the code already works, most of it driven by default java behavior, and I think this is the correct way to do it.
 
- when no context language is set
  - use local language if defined (base translation is used for undefined keys)
  - else use base translation only
- when context language is set
  - use specified language if defined (base translation is used for undefined keys)
  - else use local language if defined (base translation is used for undefined keys)
  - else use base translation only

The unset context language scenario displays English messages when the local language is English because it falls back to the base translation, the same as any other untranslated language. French is selected when the local language is French.

The scenario where the context language is set lets you specify French when the local language is English, but it does not let you specify English when the local language is French, due to the missing file provided by this PR.

File: testsrc/org/mozilla/javascript/tests/MessageProviderLanguageTest.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        assertEquals(englishMessage, getMessageUsingContextLocale(US, NOT_TRANSLATED));
+    }
+
+    @Test
+    public void defaultLocaleIsNotTranslatedAndContextLocaleUsesOtherTranslatedLanguage() {
+        assertEquals(frenchMessage, getMessageUsingContextLocale(FRANCE, NOT_TRANSLATED));
+    }
+
+    private String getMessageUsingDefaultLocale(Locale defaultLocale) {
+        Locale.setDefault(defaultLocale);
```
Comment (rbri) [inline testsrc/org/mozilla/javascript/tests/MessageProviderLanguageTest.java:104]: I think a test should not change the system status. What about saving the current locale and restore in a finally block.

**Comment by rbri** [comment]
Personally I see no reason for providing non English error messages. Maybe being English only is the simplest solution here.

File: testsrc/org/mozilla/javascript/tests/MessageProviderLanguageTest.java (hunk 1 above)
Comment (gbrail) [reply testsrc/org/mozilla/javascript/tests/MessageProviderLanguageTest.java:104]: Yes, thanks! We already have a number of tests that don't clean up the environment properly so we have to fork them into separate JVMs, which means that the test run takes much longer.

**Comment by rbri** [comment]
Any news here? Should we do a poll? ;-)

**Comment by rbri** [comment]
Have merged this because
* this is an improvement for the current situation
* we can still go on with more or less languages

**Comment by rbri** [comment]
and thanks for the PR

**Pull request merged**